src/make-twig-include.js

```diff
@@ -1,13 +1,80 @@
+/**
+ * Convert an object of story arguments into a Twig `with` string for use in
+ * template source strings.
+ * @param {any} args
+ */
+export const makeArgsString = (args = {}) => {
+  if (Object.keys(args).length > 0) {
+    return ` with ${JSON.stringify(args, null, 2)}`;
+  }
+
+  return '';
+};
+
 /**
  * Generate a twig source string of an include with args, i.e.
  * {% include '@cloudfour/components/button.twig' with { "type": "disabled" } only %}
  * @param {string} path
  * @param {any} args
  */
 export const makeTwigInclude = (path, args = {}) => {
-  const argsString =
-    Object.keys(args).length > 0
-      ? ` with ${JSON.stringify(args, null, 2)}`
-      : '';
+  const argsString = makeArgsString(args);
   return `{% include '${path}'${argsString} only %}`;
 };
+
+/**
+ * Generate a twig source string of an embed with args and blocks. If block
+ * names are included in args, their content will be included as a block within
+ * the embed. If no blocks are found, an include string will be returned
+ * instead.
+ * @param {string} path
+ * @param {any} args
+ * @param {string[]} blockNames
+ */
+export const makeTwigEmbed = (path, args = {}, blockNames = []) => {
+  const blocks = [];
+  /** @type {Record<string, any>} */
+  const filteredArgs = {};
+
+  for (const [key, value] of Object.entries(args)) {
+    if (blockNames.includes(key)) {
+      blocks.push(`
+  {% block ${key} %}
+    ${value}
+  {% endblock %}`);
+    } else {
+      filteredArgs[key] = value;
+    }
+  }
+
+  // If no blocks are found, return an include string since an embed is not
+  // necessary.
+  if (blocks.length === 0) {
+    return makeTwigInclude(path, args);
+  }
+
+  const argsString = makeArgsString(filteredArgs);
+  return `{% embed '${path}'${argsString} only %}${blocks.join('\n')}
+{% endembed %}`;
+};
+
+/**
+ * @param {string} path
+ * @param {any} args
+ * @param {string[]} blockNames
+ */
+export const makeTwigEmbedIfHtml = (path, args = {}, blockNames = []) => {
```

Comment:
@tylersticka Are there use cases where passing a specific array of block names is required, or could we change this to loop through all the keys of args to detect which args should be blocks?

Reply:
> loop through all the keys of args to detect which args should be blocks?

Do you know a way we could do this?

Reply:
```js
  for (const [key, value] of Object.keys(args)) {
    if (
      typeof value === 'string' &&
      /<\/?[a-z][\S\s]*>/i.test(value)
    ) {
      filteredBlockNames.push(value);
    }
  }
```

Reply:
The potential problem with that is that it assumes every property has an equivalent block. But most templates do not. And sometimes the block and the property aren't equivalent. For example:

```twig
<header>
  {% block header %}
    <h1>{{heading}}</h1>
  {% endblock %}
</header>
```

Reply:
I'm not sure I understand how properties and blocks work together 🤔 

If a property has HTML in it, are there cases where it will need to just be passed as a prop and not as a block?

Reply:
Props and blocks are different things. It's sort of like with web components the difference between an attribute and a slot.

In this case, they happen to be equivalent because I wanted someone to be able to use a simple property _or_ use a block for markup...

```twig
{# The template is written like this: #}
{% block content %}
  {{content}}
{% endblock %}

{# Which means these two things are equivalent: #}
{% include 'path/to/example.twig' with { content: 'foo' } %}
{% embed 'path/to/example.twig' %}{% block content %}foo{% endblock %}{% endembed %}
```

But there'd be nothing to stop me decoupling them completely:

```twig
<p>{{content}}</p>
<button>
  {% block content %}
    Hey there, {{firstName}} {{lastName}}!
  {% endblock %}
</button>
```

In which case those two examples would have entirely different results.